bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java

```diff
@@ -128,29 +138,74 @@ public void run()
         storageHelperExecutor.shutdown();
         try {
           serverHelperExecutor.awaitTermination(5000, TimeUnit.MILLISECONDS);
+          storageHelperExecutor.awaitTermination(5000, TimeUnit.MILLISECONDS);
         } catch (InterruptedException ex) {
           logger.debug("Executor Termination", ex);
         }
         logger.info("Server stopped listening at {}", address);
+        try {
+          lock.lock();
+          address = null;
+          condition.signal();
+        } finally {
+          lock.unlock();
+        }
       }
     });
   }
 
-  public synchronized InetSocketAddress run(EventLoop eventloop)
+  public InetSocketAddress run(EventLoop eventloop)
+  {
+    return run(eventloop, 0);
+  }
+
+  public InetSocketAddress run(EventLoop eventloop, long time)
   {
     eventloop.start(null, port, this);
-    while (address == null) {
-      try {
-        wait(20);
-      } catch (InterruptedException ex) {
-        throw new RuntimeException(ex);
+    long endTime = time > 0 ? System.currentTimeMillis() + time : Long.MAX_VALUE;
```

Comment:
That is practically speaking, isn't it? I understand it will be uniform but prefer not to limit it by possibly doing something like (time == 0) || (System.currentTimeMillis() < endTime). It also hinges on my other question on condition.await(0..) being indefinite wait, I couldn't find it documented anywhere. If that is true then this is ok.